pkg/controller/meta_pod_controller.go

```diff
@@ -0,0 +1,175 @@
+
+// metaLeaderStatus sends a MetaMember request to a meta node at ip:port, determining if the node is a leader.
+func (mpc *MetaPodController) metaLeaderStatus(ctx context.Context, host string, port uint) string {
+	log := log.FromContext(ctx)
+
+	addr := fmt.Sprintf("%s:%v", host, port)
+
+	for i := 0; i < 5; i++ {
+		time.Sleep(time.Duration(i*10) * time.Millisecond)
+		conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			log.Error(err, "Unable to connect to meta pod. Retrying...")
+			continue
+		}
+		defer conn.Close()
+		c := pb.NewMetaMemberServiceClient(conn)
+
+		resp, err := c.Members(ctx, &pb.MembersRequest{})
+		if err != nil {
+			log.Error(err, "Sending MembersRequest failed")
+			continue
+		}
+
+		for _, member := range resp.Members {
+			if member.IsLeader && host == member.Address.Host && port == uint(member.Address.Port) {
+				return consts.MetaRoleLeader
+			}
+		}
+		return consts.MetaRoleFollower
+	}
+	return consts.MetaRoleUnknown
+}
+
+// getMetaPort returns the service port of the pod.
+func getMetaPort(pod *corev1.Pod) (uint, error) {
+	for _, container := range pod.Spec.Containers {
+		if container.Name == "meta" {
+			for _, containerPort := range container.Ports {
+				if containerPort.Name == "service" {
+					return uint(containerPort.ContainerPort), nil
+				}
+			}
+		}
+	}
+	return 0, fmt.Errorf("unable to retrieve the service port from pod %s", pod.ObjectMeta.Name)
+}
+
+// Reconcile handles the pods of the meta service. Will add the metaLeaderLabel to the pods.
+func (mpc *MetaPodController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	requeueInterval60s := time.Second * time.Duration(60)
+	defaultRequeueResult := ctrl.Result{RequeueAfter: requeueInterval60s}
+
+	// only reconcile when this is related to a meta pod
+	if !strings.Contains(req.Name, "meta") {
+		return defaultRequeueResult, nil
+	}
+
+	log := log.FromContext(ctx)
+
+	// get all meta pods
+	metaPods := &corev1.PodList{}
+	labels := labels.SelectorFromSet(labels.Set{"risingwave/component": "meta"})
+	listOptions := client.ListOptions{LabelSelector: labels}
+	if err := mpc.Client.List(context.Background(), metaPods, &listOptions); err != nil {
+		log.Error(err, "unable to fetch meta pods")
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Do not requeue, since we do not have any meta pods
+	if len(metaPods.Items) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	hasUnknown := false
+	hasLeader := false
+	for _, pod := range metaPods.Items {
+		podIp := pod.Status.PodIP
+		port, err := getMetaPort(&pod)
+		if err != nil {
+			log.Error(err, "Error. Retrying...")
+			return defaultRequeueResult, err
+		}
+
+		// set meta label
+		originalPod := pod.DeepCopy()
+		oldRole, ok := pod.Labels[consts.LabelRisingWaveMetaRole]
+		timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Duration(3)*time.Second)
```

Comment:
As I described above, cascading the cancellation would be better.
```suggestion
		timeoutCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
```